datafusion/core/src/execution/context.rs

```diff
@@ -1290,109 +1228,88 @@ impl SessionConfig {
-        self.config_options
-            .set_bool(OPT_REPARTITION_AGGREGATIONS, enabled);
+        self.options.built_in.optimizer.repartition_aggregations = enabled;
         self
     }
 
     /// Enables or disables the use of repartitioning for window functions to improve parallelism
     pub fn with_repartition_windows(mut self, enabled: bool) -> Self {
-        self.config_options
-            .set_bool(OPT_REPARTITION_WINDOWS, enabled);
+        self.options.built_in.optimizer.repartition_windows = enabled;
         self
     }
 
     /// Enables or disables the use of pruning predicate for parquet readers to skip row groups
     pub fn with_parquet_pruning(mut self, enabled: bool) -> Self {
-        self.config_options
-            .set_bool(OPT_PARQUET_ENABLE_PRUNING, enabled);
+        self.options.built_in.execution.parquet.pruning = enabled;
         self
     }
 
     /// Returns true if pruning predicate should be used to skip parquet row groups
     pub fn parquet_pruning(&self) -> bool {
-        self.config_options
-            .get_bool(OPT_PARQUET_ENABLE_PRUNING)
-            .unwrap_or(false)
+        self.options.built_in.execution.parquet.pruning
     }
 
     /// Enables or disables the collection of statistics after listing files
     pub fn with_collect_statistics(mut self, enabled: bool) -> Self {
-        self.config_options
-            .set_bool(OPT_COLLECT_STATISTICS, enabled);
+        self.options.built_in.execution.collect_statistics = enabled;
         self
     }
 
     /// Get the currently configured batch size
     pub fn batch_size(&self) -> usize {
-        self.config_options
-            .get_u64(OPT_BATCH_SIZE)
-            .unwrap_or_default()
-            .try_into()
-            .unwrap()
-    }
-
-    /// Enables or disables the coalescence of small batches into larger batches
-    pub fn with_coalesce_batches(mut self, enabled: bool) -> Self {
-        self.config_options.set_bool(OPT_COALESCE_BATCHES, enabled);
-        self
-    }
-
-    /// Returns true if record batches will be examined between each operator
-    /// and small batches will be coalesced into larger batches.
-    pub fn coalesce_batches(&self) -> bool {
-        self.config_options
-            .get_bool(OPT_COALESCE_BATCHES)
-            .unwrap_or_default()
+        self.options.built_in.execution.batch_size
     }
 
-    /// Enables or disables the round robin repartition for increasing parallelism
-    pub fn with_round_robin_repartition(mut self, enabled: bool) -> Self {
-        self.config_options
-            .set_bool(OPT_ENABLE_ROUND_ROBIN_REPARTITION, enabled);
-        self
-    }
+    /// Convert configuration options to name-value pairs with values
+    /// converted to strings.
+    ///
+    /// Note that this method will eventually be deprecated and
+    /// replaced by [`config_options`].
+    ///
+    /// [`config_options`]: SessionContext::config_option
+    pub fn to_props(&self) -> HashMap<String, String> {
```

Comment:
This may be related to https://github.com/apache/arrow-datafusion/pull/4758 from @yahoNanJing